### PR TITLE
Update GPU SHAP for kmeans sampling code from cuML

### PR DESCRIPTION
**Pull request by Nanthini10**
cuML library added the support for kmeans sampling, so adding the support for version 21.08 (still supports previous versions)

### CODE REVIEW

**Review by imatiach-msft — APPROVED** [review]
Approved!

File: python/interpret_community/common/gpu_kmeans.py [hunk 1]
```diff
@@ -50,6 +51,15 @@ def kmeans(X, k, round_values=True):
         raise RuntimeError(
             "cuML is required to use GPU explainers. Check https://rapids.ai/start.html \
             for more information on how to install it.")
+    if cuml.__version__ >= '21.06':
+        from cuml.explainer.sampling import kmeans_sampling
+        summary, group_names, labels = kmeans_sampling(X, k, round_values, detailed=True)
+
+        return DenseData(summary,
```
Comment (imatiach-msft) [inline python/interpret_community/common/gpu_kmeans.py:58]: hmm I wish there was some way to add tests for these to ensure they are working correctly, maybe I need to look into getting GPU VMs for running cuML functionality

Comment (Nanthini10) [reply python/interpret_community/common/gpu_kmeans.py:58]: Adding versioned tests in pytest might help, but yes a second set of eyes to double check it works would be really useful!

**Comment by imatiach-msft** [comment]
@Nanthini10 can you accept the DCO so I can merge?

image

**Comment by Nanthini10** [comment]
@imatiach-msft Oh my bad, done now!

**Comment by imatiach-msft** [comment]
@Nanthini10  Merged, thank you for this contribution!  It will be in the next interpret-community release, which should go out in the next two weeks or so.

**Comment by imatiach-msft** [comment]
@Nanthini10 I tried to send you an email, not sure if you received it, about debugging running GPUTreeExplainer on an NC6 instance in AzureML (I am trying to add a notebook to show how to run GPUTreeExplainer on an AMLCompute cluster).  I am getting an error though, it can't seem to find the cuda driver.  Let me know if you have any ideas on how I might debug this issue or if anyone at NVidia might be able to help.  I sent a jupyter notebook that will eventually go in our docs here:

https://github.com/Azure/MachineLearningNotebooks/tree/master/how-to-use-azureml/explain-model/azure-integration

to show users how to run cuML GPUTreeExplainer in AzureML.

**Comment by Nanthini10** [comment]
@imatiach-msft I did look at it briefly. I'm going to debug it today, will keep you posted.

**Comment by Nanthini10** [comment]
@imatiach-msft It doesn't let me view the `gpu_tree_explainer.py` file due to security issues. Can you share that here?

**Comment by imatiach-msft** [comment]
[gpu_tree_explainer_py.txt](https://github.com/interpretml/interpret-community/files/7001420/gpu_tree_explainer_py.txt)
sure, just change the extension back to .py instead of .txt

**Comment by imatiach-msft** [comment]
tagging @Nanthini10 
https://github.com/interpretml/interpret-community/files/7001420/gpu_tree_explainer_py.txt
sure, just change the extension back to .py instead of .txt

**Comment by Nanthini10** [comment]
Thanks, got the file now! 👍🏽

**Comment by Nanthini10** [comment]
The error is coming from CUDA_PATH not being set, I'll see if I can work out a solution for this.

In the meanwhile, xgboost>=1.3 supports GPUTreeSHAP in the library (uses RAPIDS underneath). 
https://medium.com/rapids-ai/gpu-accelerated-shap-values-with-xgboost-1-3-and-rapids-587fad6822 

You should be able to get SHAP values as follows:
```
# Make sure GPU prediction is enabled
# Set "cpu_predictor" to use CPU
bst.set_param({"predictor": "gpu_predictor"})

# Compute our shap values
start = time.time()
shap_values = bst.predict(dmat, pred_contribs=True)
```

See docs: https://xgboost.readthedocs.io/en/latest/gpu/index.html

**Comment by imatiach-msft** [comment]
@Nanthini10 awesome!  How can I set the cuda path?  I thought it should be set on that Nvidia image already.

**Comment by imatiach-msft** [comment]
I tried several images including this one in the notebook:
'mcr.microsoft.com/azureml/openmpi4.1.0-cuda11.0.3-cudnn8-ubuntu18.04'

**Comment by imatiach-msft** [comment]
the images come from here:
https://github.com/Azure/AzureML-Containers
and supposedly should have everything setup for cuda already, so I'm surprised it's not working.
I did try several different images noted there, but they all had the same issue.

**Comment by imatiach-msft** [comment]
in the dockerfile you can see all of these images use the nvidia base images:

https://github.com/Azure/AzureML-Containers/blob/master/base/gpu/openmpi4.1.0-cuda11.1-cudnn8-ubuntu18.04/Dockerfile

```
FROM nvidia/cuda:11.1.1-cudnn8-devel-ubuntu18.04
```
link:

https://github.com/Azure/AzureML-Containers/blob/master/base/gpu/openmpi4.1.0-cuda11.1-cudnn8-ubuntu18.04/Dockerfile#L11

**Comment by Nanthini10** [comment]
It's possibly how SHAP is interacting with it. When I 
`echo $CUDA_PATH` it returns a blank. Other libraries might not be relying specifically on this path, but SHAP explicitly states it needs it: https://shap.readthedocs.io/en/latest/generated/shap.explainers.GPUTree.html 

I'm seeing where cuda is installed, apologies for the delay my internet is acting up and the VPN connection is super unstable.

**Comment by Nanthini10** [comment]
It's possibly how SHAP is interacting with it. When I 
`echo $CUDA_PATH` it returns a blank. Other libraries might not be relying specifically on this path, but SHAP explicitly states it needs it: https://shap.readthedocs.io/en/latest/generated/shap.explainers.GPUTree.html 

I'm seeing where cuda is installed, apologies for the delay my internet is acting up and the VPN connection is super unstable.

**Comment by Nanthini10** [comment]
This open issue on SHAP https://github.com/slundberg/shap/issues/1650 seems to be addressing this. But even with `export CUDA_PATH=/usr/local/cuda` it doesn't work and yields the same error.

**Comment by imatiach-msft** [comment]
@Nanthini10 interesting, thank you for the link to that issue, is there some other image I could use there to make it work?  I don't have to use any of those images, I just wanted to get this to work end to end.  Even creating a custom image or modifying those images to have that CUDA_PATH set correctly would be a great solution.

**Comment by Nanthini10** [comment]
You could use any of the RAPIDS image base dockerfile.

This is what we do in our [demo notebook](https://github.com/rapidsai/cloud-ml-examples/blob/main/azure/notebooks/remote-explanation/azure-gpu-shap.ipynb)

```
from azureml.core import Environment

environment_name = "rapids"

env = Environment(environment_name)
env.docker.enabled = True
env.docker.base_image = None
#Installing interpret-community from source for now, will update later
env.docker.base_dockerfile = """
FROM rapidsai/rapidsai:21.06-cuda11.0-runtime-ubuntu18.04-py3.8
RUN apt-get update && \
apt-get install -y fuse && \
apt-get install -y build-essential && \
apt-get install -y python3-dev && \
source activate rapids && \
pip install azureml-defaults && \
pip install shap==0.39
"""
env.python.user_managed_dependencies = True
```

**Comment by imatiach-msft** [comment]
@Nanthini10 thank you, I will try that!

**Comment by imatiach-msft** [comment]
actually that's an amazing example notebook:
https://github.com/rapidsai/cloud-ml-examples/tree/main/azure/notebooks/remote-explanation
I think we should have some links pointing to that in our docs for customers

**Comment by Nanthini10** [comment]
That would be great, if you can include that! Feel free to do so. We also have a blog explaining the whole integration for those interested in the technical details: https://developer.nvidia.com/blog/model-interpretability-using-rapids-implementation-of-shap-on-microsoft-azure/

**Comment by imatiach-msft** [comment]
@Nanthini10 that's an amazing blog, thanks for the link!  We will be sure to link to it as well.  I'm discussing with our PMs currently where we could add relevant links.

**Comment by imatiach-msft** [comment]
@Nanthini10 I tried to run the notebook:

https://github.com/rapidsai/cloud-ml-examples/blob/main/azure/notebooks/remote-explanation/azure-gpu-shap.ipynb

But I got the error:

Traceback (most recent call last):
  File "/rapids/src/interpret-community/python/interpret_community/tabular_explainer.py", line 177, in __init__
    self.explainer = uninitialized_explainer(
  File "/rapids/src/interpret-community/python/interpret_community/dataset/decorator.py", line 42, in init_wrapper
    return init_func(self, model, wrap_dataset(initialization_examples), *args, **kwargs)
  File "/rapids/src/interpret-community/python/interpret_community/common/blackbox_explainer.py", line 183, in init_wrapper
    return init_func(self, model, *args, **kwargs)
  File "/rapids/src/interpret-community/python/interpret_community/common/aggregate.py", line 24, in init_wrapper
    return init_func(self, model, *args, **kwargs)
  File "/rapids/src/interpret-community/python/interpret_community/shap/gpu_kernel_explainer.py", line 176, in __init__
    self._reset_evaluation_background(self.function, **kwargs)
  File "/rapids/src/interpret-community/python/interpret_community/shap/gpu_kernel_explainer.py", line 187, in _reset_evaluation_background
    function, summary = self._prepare_function_and_summary(function, self.original_data_ref,
  File "/rapids/src/interpret-community/python/interpret_community/common/blackbox_explainer.py", line 244, in _prepare_function_and_summary
    self.initialization_examples.compute_summary(**kwargs)
  File "/rapids/src/interpret-community/python/interpret_community/dataset/dataset_wrapper.py", line 451, in compute_summary
    self._summary_dataset = _summarize_data(self._dataset, nclusters, use_gpu)
  File "/rapids/src/interpret-community/python/interpret_community/common/explanation_utils.py", line 61, in _summarize_data
    return kmeans(X, k, to_round_values)
  File "/rapids/src/interpret-community/python/interpret_community/common/gpu_kmeans.py", line 55, in kmeans
    from cuml.explainer.sampling import kmeans_sampling
ModuleNotFoundError: No module named 'cuml.explainer.sampling'


It looks like it is related to this PR?

**Comment by imatiach-msft** [comment]
Also I figured out that I needed to set my cluster to use:

Standard_NC6s_v3

Using the VMs:

Standard_NC6

used the older K80s which apparently doesn't work with the rapids image? but these seem to work apparently:
GPU - 1 x NVIDIA Tesla V100

**Comment by Nanthini10** [comment]
@imatiach-msft Ahh I see the issue, it should be cuml version >=21.08 in this PR. Can you try updating the notebook to pull `rapidsai/rapidsai:21.08-cuda11.0-runtime-ubuntu18.04-py3.8` and try? I'll make another PR to fix this issue

Yeah, RAPIDS works only with Pascal or better GPUs. In case of Azure, you'd need the v3 GPUs to run the notebook.

**Comment by Nanthini10** [comment]
I haven't had a chance to update the notebook since the merge that's why it uses the older versions of both libraries.

**Comment by Nanthini10** [comment]
@imatiach-msft Updated https://github.com/interpretml/interpret-community/pull/436

I thought I checked on both version must have missed it, sorry about that!

**Comment by imatiach-msft** [comment]
@Nanthini10 interesting, I tried the new image but I got a new error:

Failure while loading interpret_ext_blackbox. Failed to load entrypoint KernelExplainer = interpret_community.shap:KernelExplainer with exception Traceback (most recent call last):
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/interpret/ext/extension_utils.py", line 45, in load_class_extensions
    extension_class = entrypoint.load()
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2472, in load
    return self.resolve()
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2478, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/rapids/notebooks/src/interpret-community/python/interpret_community/__init__.py", line 10, in <module>
    from .tabular_explainer import TabularExplainer
  File "/rapids/notebooks/src/interpret-community/python/interpret_community/tabular_explainer.py", line 11, in <module>
    from .shap.tree_explainer import TreeExplainer
  File "/rapids/notebooks/src/interpret-community/python/interpret_community/shap/__init__.py", line 10, in <module>
    from .gpu_kernel_explainer import GPUKernelExplainer
  File "/rapids/notebooks/src/interpret-community/python/interpret_community/shap/gpu_kernel_explainer.py", line 39, in <module>
    import cuml
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cuml/__init__.py", line 17, in <module>
    from cuml.common.base import Base
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cuml/common/__init__.py", line 17, in <module>
    from cuml.common.array import CumlArray
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cuml/common/array.py", line 23, in <module>
    from cudf.core import Buffer, Series, DataFrame
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/__init__.py", line 11, in <module>
    from cudf import core, datasets, testing
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/__init__.py", line 3, in <module>
    from cudf.core import _internals, buffer, column, column_accessor, common
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/_internals/__init__.py", line 3, in <module>
    from cudf.core._internals.where import where
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/_internals/where.py", line 11, in <module>
    from cudf.core.column import ColumnBase
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/column/__init__.py", line 7, in <module>
    from cudf.core.column.categorical import CategoricalColumn
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/column/categorical.py", line 24, in <module>
    from cudf import _lib as libcudf
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/_lib/__init__.py", line 4, in <module>
    from . import (
  File "cudf/_lib/avro.pyx", line 1, in init cudf._lib.avro
  File "cudf/_lib/table.pyx", line 1, in init cudf._lib.table
  File "cudf/_lib/column.pyx", line 1, in init cudf._lib.column
  File "cudf/_lib/scalar.pyx", line 36, in init cudf._lib.scalar
  File "cudf/_lib/interop.pyx", line 1, in init cudf._lib.interop
AttributeError: module 'pyarrow.lib' has no attribute 'StopToken'

let me try master again, maybe it will work now that this PR has been merged

**Comment by Nanthini10** [comment]
That's interesting, considering that I used the same image locally for testing (I used Python 3.8 instead of 3.7 but I don't think that is the issue here), it should be running.


I'm caught up with some other things today and will have a chance to update the notebook sometime this week.

**Comment by imatiach-msft** [comment]
"I used Python 3.8 instead of 3.7 but I don't think that is the issue here"
interesting, that could be the issue, do I just need to change the image name to:

rapidsai/rapidsai-core:21.08-cuda11.0-runtime-ubuntu18.04-py3.8

**Comment by imatiach-msft** [comment]
are you sure "rapidsai-core" is the right one and not just rapidsai like the original one?

**Comment by Nanthini10** [comment]
> @Nanthini10
> "I used Python 3.8 instead of 3.7 but I don't think that is the issue here"
> interesting, that could be the issue, do I just need to change the image name to:
> 
> rapidsai/rapidsai-core:21.08-cuda11.0-runtime-ubuntu18.04-py3.8

Maybe so. Yeah `rapidsai/rapidsai-core:21.08-cuda11.0-runtime-ubuntu18.04-py3.8` is the image

Yes, `rapidsai` hasn't been updated yet. The core just ignores the `blazing` library, so it shouldn't affect us in any way.

**Comment by imatiach-msft** [comment]
@Nanthini10 I can confirm that using the old image:
```
FROM rapidsai/rapidsai:21.06-cuda11.0-runtime-ubuntu18.04-py3.8
```

with the latest interpret-community from master with your recent checkin finally worked!! Thanks a lot for helping me with this!
Unfortunately I am still trying to figure out how to get shap GPUTreeExplainer to work.

This is the latest dockerfile I am using (note with just shap and no interpret-community):

env.docker.base_dockerfile = """
FROM rapidsai/rapidsai:21.06-cuda11.0-runtime-ubuntu18.04-py3.8
RUN apt-get update && \
apt-get install -y fuse && \
apt-get install -y build-essential && \
apt-get install -y python3-dev && \
source activate rapids && \
apt-get install -y g++ && \
pip install azureml-defaults && \
pip install azureml-telemetry && \
pip install git+git://github.com/slundberg/shap@master && \
pip install xgboost==1.4.2
"""

but I am getting the error:
```
Traceback (most recent call last):
  File "gpu_tree_explainer.py", line 27, in <module>
    shap_values = explainer(X_shap)
  File "/opt/conda/envs/rapids/lib/python3.8/site-packages/shap/explainers/_tree.py", line 211, in __call__
    v = self.shap_values(X, y=y, from_call=True, check_additivity=check_additivity)
  File "/opt/conda/envs/rapids/lib/python3.8/site-packages/shap/explainers/_gpu_tree.py", line 105, in shap_values
    assert_import("cext_gpu")
  File "/opt/conda/envs/rapids/lib/python3.8/site-packages/shap/utils/_general.py", line 23, in assert_import
    raise e
  File "/opt/conda/envs/rapids/lib/python3.8/site-packages/shap/explainers/_gpu_tree.py", line 6, in <module>
    from .. import _cext_gpu
ImportError: cannot import name '_cext_gpu' from partially initialized module 'shap' (most likely due to a circular import) (/opt/conda/envs/rapids/lib/python3.8/site-packages/shap/__init__.py)
```

**Comment by Nanthini10** [comment]
@imatiach-msft That's great! Thanks for the update!

Unfortunately, that error seems to be coming from SHAP - it's the way they set up the GPUTreeExplainer. 
In case of RAPIDS containers `CUDA_PATH` is already set, so I'm quite unsure what is causing this.

**Comment by imatiach-msft** [comment]
@Nanthini10 I managed to get the GPUTree explainer to work with the image:

env.docker.base_dockerfile = """
FROM mirekphd/ml-gpu-py38-cuda101-cust
RUN pip install azureml-defaults && \
pip install azureml-telemetry && \
pip install xgboost==1.4.2
"""

as recommended in this thread:

https://github.com/slundberg/shap/issues/1650

but I would prefer to use an official nvidia image instead of one created by a user unofficially which might not be supported in the future

**Comment by Nanthini10** [comment]
@imatiach-msft That's great! Since the problem is coming from shap, there's little we can do with the official NVIDIA images. We can run RAPIDS GPUTreeExplainer as is used by xgboost library, but this fix should come from shap unfortunately.

**Pull request merged**